Sources/PostgresNIO/Connection/PostgresConnection+Notifications.swift

```diff
@@ -16,47 +16,65 @@ extension PostgresConnection {
     /// Add a handler for NotificationResponse messages on a certain channel. This is used in conjunction with PostgreSQL's `LISTEN`/`NOTIFY` support: to listen on a channel, you add a listener using this method to handle the NotificationResponse messages, then issue a `LISTEN` query to instruct PostgreSQL to begin sending NotificationResponse messages.
     @discardableResult
     public func addListener(channel: String, handler notificationHandler: @escaping (PostgresListenContext, PostgresMessage.NotificationResponse) -> Void) -> PostgresListenContext {
+        
         let listenContext = PostgresListenContext()
-        let channelHandler = PostgresNotificationHandler(logger: self.logger, channel: channel, notificationHandler: notificationHandler, listenContext: listenContext)
-        let pipeline = self.channel.pipeline
-        _ = pipeline.addHandler(channelHandler, name: nil, position: .last)
-        listenContext.stopper = { [pipeline, unowned channelHandler] in
-            _ = pipeline.removeHandler(channelHandler)
+        
+        self.underlying.channel.pipeline.handler(type: PSQLChannelHandler.self).whenSuccess { handler in
+            if self.notificationListeners[channel] != nil {
+                self.notificationListeners[channel]!.append((listenContext, notificationHandler))
+            }
+            else {
+                self.notificationListeners[channel] = [(listenContext, notificationHandler)]
+            }
+        }
+        
+        listenContext.stopper = { [weak self, weak listenContext] in
+            // self is weak, since the connection can long be gone, when the listeners stop is
+            // triggered. listenContext must be weak to prevent a retain cycle
+            
+            self?.underlying.channel.eventLoop.execute {
+                guard let self = self else {
+                    // the connection is already gone
+                    return
+                }
+                
+                guard var listeners = self.notificationListeners[channel] else {
+                    // we don't have the listeners for this topic ¯\_(ツ)_/¯
+                    return
+                }
+                
+                guard let index = listeners.firstIndex(where: { $0.0 === listenContext }) else {
+                    return
+                }
+                
+                listeners.remove(at: index)
+                if listeners.count == 0 {
+                    self.notificationListeners.removeValue(forKey: channel)
+                } else {
+                    self.notificationListeners[channel] = listeners
+                }
```

Comment:
Yes, that looks really smart, but I'm not really sure, if this should be a goal of the code here... And since we write a low level driver IMHO we should not leak objects, just to cram everything into one line.

Reply:
I agree, to some extent. May I suggest this compromise as an alternative?
```suggestion
                guard
                    let self = self, // the connection is already gone
                    var listeners = self.notificationListeners[channel] // we don't have the listeners for this topic ¯\_(ツ)_/¯
                else {
                    return
                }
                
                assert(listeners.filter { $0.0 === listenContext }.count <= 1, "Listeners can not appear twice in a channel!")
                listeners.removeAll(where: { $0.0 === listenContext }) // just in case a listener shows up more than once in a release build, remove all, not just first
                self.notificationListeners[channel] = listeners.isEmpty ? nil : listeners
```